### PR TITLE
"tsc -b" with minimal watch capabilities

### PR DESCRIPTION
**Pull request by RyanCavanaugh**
See #22997

Implements `tsbuild` without `--watch`. Opening so people have a chance to review non-watch portions sooner. Not planning to merge this unless there is schedule pressure to produce a non-watch build.

I also want to rephrase ~all of the error messages and would like suggestions on any of them

@rbuckton there's a vfs fix in here because we were re-using `SourceFile` instances even when the underlying file was changed

Open TODOs I'm aware of:
 * [ ] Nail down which parts of the tsbuilder API should be exposed (seeking feedback)
 * [x] Determine exactly what `--verbose` output should be in all situations (seeking feedback)
 * [x] Issue an error when `-b` appears out of order in a commandline
 * [x] Issue an error when an illegal set of `-b` flags is provided (e.g. `--clean --dry --force`, which makes no sense)
 * [x] Add an `outFile` compilation to the sample projects folder
 * [x] `--build --help`
 * [ ] Support response files in `--build` mode (do we need this?)

### CODE REVIEW

File: src/harness/vfs.ts [hunk 1]
```diff
@@ -404,7 +409,21 @@ namespace vfs {
         }
 
         /**
-         * Get file status.
+         * Change file access times
+         *
+         * NOTE: do not rename this method as it is intended to align with the same named export of the "fs" module.
+         */
+        public utimesSync(path: string, atime: Date, mtime: Date) {
```
Comment (rbuckton) [inline src/harness/vfs.ts]: I originally had this as part of the VFS, the relevant code can be found here: https://gist.github.com/rbuckton/1d064ec108afc0479bc1457d8a46792e#file-filesystem-ts-L596

You should verify the file system isn't read-only before performing updates so that you don't inadvertently modify a copy of the FS shared between multiple tests, i.e.:

```ts
if (this.isReadOnly) throw new IOError("EROFS");
```

File: src/harness/vfs.ts [hunk 2]
```diff
@@ -404,7 +409,21 @@ namespace vfs {
...
+         *
+         * NOTE: do not rename this method as it is intended to align with the same named export of the "fs" module.
+         */
+        public utimesSync(path: string, atime: Date, mtime: Date) {
+            const entry = this._walk(this._resolve(path));
+            if (!entry || !entry.node) {
+                throw createIOError("ENOENT");
+            }
+            entry.node.atimeMs = +atime;
+            entry.node.mtimeMs = +mtime;
```
Comment (rbuckton) [inline src/harness/vfs.ts]: You should also set `entry.node.ctimeMs = this.time()` to set the status change time when you update `atimeMs` or `mtimeMs`.

File: src/compiler/tsc.ts [hunk 3]
```diff
@@ -46,9 +41,25 @@ namespace ts {
...
 
+    function getOptionsForHelp(commandLine: ParsedCommandLine) {
+        // Sort our options by their names, (e.g. "--noImplicitAny" comes before "--watch")
+        return !!commandLine.options.all ?
+            sort(optionDeclarations, (a, b) => compareStringsCaseInsensitive(a.name, b.name)) :
+            filter(optionDeclarations.slice(), v => !!v.showInSimplifiedHelpView);
+    }
+
     export function executeCommandLine(args: string[]): void {
+        if ((args[0].toLowerCase() === "--build") || (args[0].toLowerCase() === "-b")) {
```
Comment (weswigham) [inline src/compiler/tsc.ts]: I'm going to reiterate what I've said before so it's written down: _I really dislike this_. This flag disables all other normal flags, must be first, and completely changes what the executable _does_ - these all indicate that it should be a _separate executable_. We can still ship it, use common code, and everything, but it should by no means be executed by `tsc --build` when it functionally shares nothing of its interface with `tsc`. A second entrypoint, `tsbuild` would make much more sense from a functionality perspective, plus it'd make the feature more googleable (and remove the awkward comments in the `CompilerOptions` interface about how "this flag should never be set"). Plus, then `tsbuild` wouldn't be forced to use `tsc`'s command line option model, and could use a much more pleasant command-based syntax (since `tsbuild` is a much more multimodal tool than `tsc`, as an orchestrator), like `tsbuild`, `tsbuild compile`, `tsbuild dryrun`, `tsbuild watch`, and so on.

File: tests/baselines/reference/outfile-concat.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    }
+    C.prototype.doSomething = function () {
+        console.log("something got done");
+    };
+    return C;
+}());
+//# sourceMappingURL=second-output.js.map
+var c = new C();
+c.doSomething();
+//# sourceMappingURL=third-output.js.map
```
Comment (weswigham) [inline tests/baselines/reference/outfile-concat.js]: We should probably avoid printing these interstitial `sourceMappingURL` comments in concatenated files. They have the potential to seriously confuse tools, should the files actually exist on disk.

Comment (weswigham) [reply tests/baselines/reference/outfile-concat.js]: Minimally, we should be emitting a bundle-wide one on the last line, to override the one from the last concatenated file for well-behaved tools.

File: src/compiler/utilities.ts [hunk 5]
```diff
@@ -2851,13 +2851,26 @@ namespace ts {
...
+
         function write(s: string) {
             if (s && s.length) {
                 if (lineStart) {
-                    output += getIndentString(indent);
+                    s = getIndentString(indent) + s;
                     lineStart = false;
                 }
                 output += s;
+                updateLineCountAndPosFor(s);
```
Comment (mhegazy) [inline src/compiler/utilities.ts]: So we are walking now every string we write in the emitter to find lineStarts? why is this needed? do you know the perf impact on the emitter?

File: src/compiler/tsbuild.ts [hunk 6]
```diff
@@ -0,0 +1,1222 @@
...
+
+        /**
+         * Map from config file name to up-to-date status
+         */
+        projectStatus: FileMap<UpToDateStatus>;
+
+        /**
+         * Issue a verbose diagnostic message. No-ops when options.verbose is false.
+         */
+        verbose(diag: DiagnosticMessage, ...args: any[]): void;
```
Comment (mhegazy) [inline src/compiler/tsbuild.ts]: in other parts of the system, we have done this through host objects.

File: src/compiler/tsbuild.ts [hunk 7]
```diff
@@ -0,0 +1,1222 @@
...
+        removeKey(fileName: string): void;
+        getKeys(): string[];
+    }
+
+    /**
+     * A FileMap maintains a normalized-key to value relationship
+     */
+    function createFileMap<T>(): FileMap<T> {
+        // tslint:disable-next-line:no-null-keyword
+        const lookup: { [key: string]: T } = Object.create(/*prototype*/ null);
```
Comment (mhegazy) [inline src/compiler/tsbuild.ts]: createMap?

Comment (mhegazy) [reply src/compiler/tsbuild.ts]: ES Maps has better memory characteristics than objects, since they do not hold on to the property name strings.

File: src/compiler/tsbuild.ts [hunk 8]
```diff
@@ -0,0 +1,1222 @@
...
+        }
+
+        function addEntry(mapToAddTo: typeof childToParents | typeof parentToChildren, key: ResolvedConfigFileName, element: ResolvedConfigFileName) {
+            key = normalizePath(key) as ResolvedConfigFileName;
+            element = normalizePath(element) as ResolvedConfigFileName;
+            const arr = (mapToAddTo[key] = mapToAddTo[key] || []);
+            if (arr.indexOf(element) < 0) {
+                arr.push(element);
+            }
+            if (allKeys.indexOf(key) < 0) allKeys.push(key);
```
Comment (mhegazy) [inline src/compiler/tsbuild.ts]: consider using a map and returning keys, instead of using `indexof`

File: src/compiler/tsbuild.ts [hunk 9]
```diff
@@ -0,0 +1,1222 @@
...
+            addReference,
+            getReferencesTo,
+            getReferencesOf,
+            getKeys
+        };
+    }
+
+    function getOutputDeclarationFileName(inputFileName: string, configFile: ParsedCommandLine) {
+        const relativePath = getRelativePathFromDirectory(rootDirOfOptions(configFile.options, configFile.options.configFilePath!), inputFileName, /*ignoreCase*/ true);
+        const outputPath = resolvePath(configFile.options.declarationDir || configFile.options.outDir || getDirectoryPath(configFile.options.configFilePath!), relativePath);
```
Comment (mhegazy) [inline src/compiler/tsbuild.ts]: program has a similar function, can we consolidate?

File: src/compiler/tsbuild.ts [hunk 10]
```diff
@@ -0,0 +1,1222 @@
...
+        return changeExtension(outputPath, ".d.ts");
+    }
+
+    function getOutputJavaScriptFileName(inputFileName: string, configFile: ParsedCommandLine) {
```
Comment (mhegazy) [inline src/compiler/tsbuild.ts]: looks like getOwnEmitOutputFilePath. can we consolidate?

File: src/compiler/tsbuild.ts [hunk 11]
```diff
@@ -0,0 +1,1222 @@
...
+        const relativePath = getRelativePathFromDirectory(rootDirOfOptions(configFile.options, configFile.options.configFilePath!), inputFileName, /*ignoreCase*/ true);
+        const outputPath = resolvePath(configFile.options.outDir || getDirectoryPath(configFile.options.configFilePath!), relativePath);
+        return changeExtension(outputPath, (fileExtensionIs(inputFileName, ".tsx") && configFile.options.jsx === JsxEmit.Preserve) ? ".jsx" : ".js");
+    }
+
+    function getOutputFileNames(inputFileName: string, configFile: ParsedCommandLine): ReadonlyArray<string> {
+        if (configFile.options.outFile) {
```
Comment (mhegazy) [inline src/compiler/tsbuild.ts]: there is `out` as well.

File: src/compiler/utilities.ts (hunk 5 above)
Comment (weswigham) [reply src/compiler/utilities.ts]: If we don't count the lines and the length of the last one, we don't set the line and column correctly for get line and column, resulting in incorrect sourcemaps.

Comment (mhegazy) [reply src/compiler/utilities.ts]: i understand, but how often are these functions called with multi-line text? can we split these to have their own writer function to avoid doing this on every token we write?

File: src/compiler/tsbuild.ts [hunk 12]
```diff
@@ -0,0 +1,1222 @@
...
+            if (configFile.options.declarationMap) {
+                outputs.push(dts + ".map");
+            }
+        }
+        return outputs;
+    }
+
+    function getOutFileOutputs(project: ParsedCommandLine): ReadonlyArray<string> {
+        if (!project.options.outFile) {
+            throw new Error("Assert - outFile must be set");
```
Comment (mhegazy) [inline src/compiler/tsbuild.ts]: `Debug.fail`

File: src/compiler/tsbuild.ts [hunk 13]
```diff
@@ -0,0 +1,1222 @@
...
+        }
+        const outputs: string[] = [];
+        outputs.push(project.options.outFile);
```
Comment (mhegazy) [inline src/compiler/tsbuild.ts]: this looks a lot like getOutputPathsFor. just use that instead.

File: src/compiler/tsbuild.ts [hunk 14]
```diff
@@ -0,0 +1,1222 @@
...
+            removeKey
+        };
+    }
+
+    function newer(date1: Date, date2: Date): Date {
+        return date2 > date1 ? date2 : date1;
+    }
+
+    function isDeclarationFile(fileName: string) {
+        return fileExtensionIs(fileName, ".d.ts");
```
Comment (mhegazy) [inline src/compiler/tsbuild.ts]: nit, use `Extension.Dts` instead of `".d.ts"`

File: src/compiler/tsbuild.ts [hunk 15]
```diff
@@ -0,0 +1,1222 @@
...
+            if (filesToDelete === undefined) {
+                reportDiagnostic(createCompilerDiagnostic(Diagnostics.Skipping_clean_because_not_all_projects_could_be_located));
+                return;
+            }
+
+            if (context.options.dry) {
+                reportDiagnostic(createCompilerDiagnostic(Diagnostics.A_non_dry_build_would_delete_the_following_files_Colon_0, filesToDelete.map(f => `\r\n * ${f}`).join("")));
+            }
+            else {
+                if (!host.deleteFile) {
```
Comment (mhegazy) [inline src/compiler/tsbuild.ts]: seems like this should be the first check

File: src/compiler/tsbuild.ts [hunk 16]
```diff
@@ -0,0 +1,1222 @@
...
+        function resolveProjectName(name: string): ResolvedConfigFileName | undefined {
+            let fullPath = resolvePath(host.getCurrentDirectory(), name);
+            if (host.fileExists(fullPath)) {
+                return fullPath as ResolvedConfigFileName;
+            }
+            fullPath = combinePaths(fullPath, "tsconfig.json");
+            if (host.fileExists(fullPath)) {
+                return fullPath as ResolvedConfigFileName;
+            }
+            reportDiagnostic(createCompilerDiagnostic(Diagnostics.File_0_not_found, fullPath));
```
Comment (mhegazy) [inline src/compiler/tsbuild.ts]: this will say something like `c:\myconfig.json\tsconfig.json` does not exist

File: src/compiler/tsbuild.ts [hunk 17]
```diff
@@ -0,0 +1,1222 @@
...
+
+            const programOptions: CreateProgramOptions = {
+                projectReferences: configFile.projectReferences,
+                host,
+                rootNames: configFile.fileNames,
+                options: configFile.options
+            };
+            const program = createProgram(programOptions);
+
+            // Don't emit anything in the presence of syntactic errors or options diagnostics
```
Comment (mhegazy) [inline src/compiler/tsbuild.ts]: why? we emit in the presence of errors in tsc.. and why are option errors different from semantic errors?

File: src/compiler/tsbuild.ts [hunk 18]
```diff
@@ -0,0 +1,1222 @@
...
+                return BuildResultFlags.None;
+            }
+
+            const programOptions: CreateProgramOptions = {
+                projectReferences: configFile.projectReferences,
+                host,
+                rootNames: configFile.fileNames,
+                options: configFile.options
+            };
+            const program = createProgram(programOptions);
```
Comment (mhegazy) [inline src/compiler/tsbuild.ts]: should not this just call inot the same routine we have in tsc.ts to perform the compilation, why does it have to be different?

File: src/compiler/tsbuild.ts [hunk 19]
```diff
@@ -0,0 +1,1222 @@
...
+                    resultFlags |= BuildResultFlags.DeclarationEmitErrors;
+                    for (const diag of declDiagnostics) {
+                        reportDiagnostic(diag);
+                    }
+                    context.projectStatus.setValue(proj, { type: UpToDateStatusType.Unbuildable, reason: "Declaration file errors" });
+                    return resultFlags;
+                }
+            }
+
+            const semanticDiagnostics = [...program.getSemanticDiagnostics()];
```
Comment (mhegazy) [inline src/compiler/tsbuild.ts]: why clone it?

File: src/compiler/tsbuild.ts (hunk 17 above)
Comment (mhegazy) [reply src/compiler/tsbuild.ts]: I see, we are not building for any errors.. nevermind then.

**Comment by mhegazy** [comment]
Overall 👍 with some comments:
- Emitter performance check. Please check the performance of the emitter did not regress from this change
- Style comments in this PR, we can merge then handle them in a different PR.

File: src/compiler/utilities.ts (hunk 5 above)
Comment (RyanCavanaugh) [reply src/compiler/utilities.ts]: Probably best to discuss this in #24704

File: src/compiler/tsbuild.ts (hunk 10 above)
Comment (RyanCavanaugh) [reply src/compiler/tsbuild.ts]: Holding off on these because `getOwnEmitOutputFilePath` needs a `SourceFile` and an `EmitHost` and we have neither at this point.

File: src/compiler/tsbuild.ts (hunk 13 above)
Comment (RyanCavanaugh) [reply src/compiler/tsbuild.ts]: Same problem here; those functions were written taking much wider inputs than required or present
